Cosim rebase

hcallahan-lowrisc opened the pull request:
Updated version of pr #1613 , as the build scripts have been significantly refactored underneath.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
+    if args.en_cosim:
+        _spike_iss_pc = 'riscv-riscv riscv-disasm riscv-fdt'
+
+        if not subprocess.run(shlex.split(f'pkg-config --exists {_spike_iss_pc}'),
```
Probably easier to make `_spike_iss_pc` a list of strings and then call something like `subprocess.run(['pkg-config', '--exists'] + _spike_iss_pc, ...)`?

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
 
+    # Setup environment variables for the tb to link against the compiled ISS
+    if args.en_cosim:
+        _spike_iss_pc = 'riscv-riscv riscv-disasm riscv-fdt'
```
Not that I feel strongly, but why the underscore prefix?

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
+
+        if not subprocess.run(shlex.split(f'pkg-config --exists {_spike_iss_pc}'),
+                              check=True):
+            raise EnvironmentError(f'Failed to find "{_spike_iss_pc}" pkg-config\
+            packages. Did you set the PKG_CONFIG_PATH correctly?')
```
I think this sort of thing is easier to read if you use Python's C-style string pasting (e.g. `"a" "b"` is the same as "ab") and stick newlines between the strings.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
+                args=shlex.split(f'pkg-config {cmd} {iss_pc}'),
+                universal_newlines=True)
+            if args.simulator == 'xlm':
+                return _flag.replace('-Wl,', '')  # Don't ask
```
I like the comment :-) But we should probably say something a little more informative for the next poor soul that has to make sense of this!

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
+            raise EnvironmentError(f'Failed to find "{_spike_iss_pc}" pkg-config\
+            packages. Did you set the PKG_CONFIG_PATH correctly?')
+
+        def _get_iss_pkgconfig_flags(cmd, iss_pc):
```
I don't think this is closing over anything? If not, I'd probably suggest moving it out of the parent function to keep everything reasonably short.

hcallahan-lowrisc replied:
This is the only comment I didn't fully address in the new changes, since it does close over args.simulator, and I like the readability of keeping the function local to its use. But if you feel strongly I don't mind moving it either.

rswarbrick replied:
Ah, I see. I'd probably prefer moving it out: that way, it's more explicit what we're passing to the helper function (and it's `args.simulator`, instead of `args`) and it also avoids the surrounding function getting too enormous.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +37,32 @@ def main() -> int:
+            'ISS_LIBS'    : '--libs-only-l --libs-only-L',
+        }
+        for k, v in _iss_pkgconfig_envvars_dict.items():
+            os.environ[k] = _get_iss_pkgconfig_flags(v, _spike_iss_pc)
```
The usual approach here is to take a copy of the environment and modify that, then pass it to subprocess (I think `run_one` has an `env` argument that you can use)

rswarbrick reviewed (commented):
Please can you squash the last commit in with the previous one (so that we don't have an intermediate state where COSIM = 1 by default)?

Also, I think some of the commit messages could be a bit more informative: once this is merged, no one will know or care about a rebase :-)

hcallahan-lowrisc commented:
Changes addressing the comments are pushed now. 
I think I've addressed everything, but stuff has moved around a little bit now. I've changed to using the subst_vars() method rather than environment variables for populating the yaml template.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +36,41 @@ def main() -> int:
+    if args.en_cosim:
+
+        try:
+            subprocess.check_output(['pkg-config'] + ['--exists'] + spike_iss_pc)
```
Nit: I'd probably suggest `['pkg-config', '--exists']` here

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +36,41 @@ def main() -> int:
+            subprocess.check_output(['pkg-config'] + ['--exists'] + spike_iss_pc)
+        except subprocess.CalledProcessError as err:
+            raise RuntimeError(
+                'Failed to find ' + str(spike_iss_pc) + ' pkg-config packages. '
+                'Did you set the PKG_CONFIG_PATH correctly?') from err
```
This is easier with format strings:
```
                f'Failed to find {spike_iss_pc} pkg-config packages. '
                f'Did you set the PKG_CONFIG_PATH correctly?') from err
```
(the second `f'...'` is optional, but I tend to do it to make everything line up!)

You also don't need the `from err` bit: that's the default behaviour.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -35,6 +36,41 @@ def main() -> int:
+        'ISS_LIBS'    : ['--libs-only-l', '--libs-only-L'],
+    }
+    if args.en_cosim:
+
```
Nit: I think the conventional layout for these things wouldn't have the newline here.

rswarbrick commented on dv/uvm/core_ibex/scripts/compile-tb.py:
```diff
@@ -49,7 +85,8 @@ def main() -> int:
                              'cmp_opts': get_compile_opts(args.ibex_config,
                                                           args.simulator)
-                         })
+                         } | cosim_subst_vars
+                         )
```
Ha! As Canberk and I discovered a couple of days ago, this operator was only added in Python 3.9. Since we only require Python 3.6, you'll have to do the ugly stateful update. (But you can, at least, hoist it out of the loop)

hcallahan-lowrisc commented:
Comments and nits addressed.
